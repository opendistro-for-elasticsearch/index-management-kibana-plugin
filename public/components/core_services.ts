import { createContext } from "react";
import { CoreStart } from "kibana/public";

const CoreServicesContext = createContext<CoreStart | null>(null);

const CoreServicesConsumer = CoreServicesContext.Consumer;

export { CoreServicesContext, CoreServicesConsumer };
