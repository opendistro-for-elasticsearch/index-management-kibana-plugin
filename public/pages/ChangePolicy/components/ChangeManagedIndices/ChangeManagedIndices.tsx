/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React, { Component } from "react";
import { EuiSpacer, EuiText, EuiComboBox, EuiFormRow } from "@elastic/eui";
import { toastNotifications } from "ui/notify";
import { ContentPanel } from "../../../../components/ContentPanel";
import { ManagedIndexService } from "../../../../services";
import { ManagedIndexItem, State } from "../../../../../models/interfaces";

interface ChangeManagedIndicesProps {
  managedIndexService: ManagedIndexService;
  selectedManagedIndices: { label: string; value?: ManagedIndexItem }[];
  selectedStateFilters: { label: string }[];
  onChangeManagedIndices: (selectedManagedIndices: { label: string; value?: ManagedIndexItem }[]) => void;
  onChangeStateFilters: (stateFilter: { label: string }[]) => void;
  managedIndicesError: string;
}

interface ChangeManagedIndicesState {
  managedIndicesIsLoading: boolean;
  managedIndices: { label: string; value?: ManagedIndexItem }[];
  stateFilterSearchValue: string;
}

export default class ChangeManagedIndices extends Component<ChangeManagedIndicesProps, ChangeManagedIndicesState> {
  state = {
    managedIndicesIsLoading: false,
    managedIndices: [],
    stateFilterSearchValue: "",
  };

  async componentDidMount(): Promise<void> {
    await this.onManagedIndexSearchChange("");
  }

  onManagedIndexSearchChange = async (searchValue: string): Promise<void> => {
    const { managedIndexService } = this.props;
    this.setState({ managedIndicesIsLoading: true, managedIndices: [] });
    try {
      // only bring back the first 10 results descending by name
      const queryParamsString = `from=0&size=10&search=${searchValue}&sortDirection=desc&sortField=name`;
      const managedIndicesResponse = await managedIndexService.getManagedIndices(queryParamsString);
      if (managedIndicesResponse.ok) {
        const options = searchValue.trim() ? [{ label: `${searchValue}*` }] : [];
        const managedIndices = managedIndicesResponse.response.managedIndices.map((managedIndex: ManagedIndexItem) => ({
          label: managedIndex.index,
          value: managedIndex,
        }));
        this.setState({ managedIndices: options.concat(managedIndices) });
      } else {
        if (managedIndicesResponse.error.startsWith("[index_not_found_exception]")) {
          toastNotifications.addDanger("You have not created a managed index yet");
        } else {
          toastNotifications.addDanger(managedIndicesResponse.error);
        }
      }
    } catch (err) {
      toastNotifications.addDanger(err.message);
    }

    this.setState({ managedIndicesIsLoading: false });
  };

  onStateFilterSearchChange = (searchValue: string): void => {
    this.setState({ stateFilterSearchValue: searchValue });
  };

  render() {
    const { managedIndices, managedIndicesIsLoading, stateFilterSearchValue } = this.state;
    const { selectedManagedIndices, selectedStateFilters, managedIndicesError } = this.props;

    const uniqueStates = selectedManagedIndices.reduce(
      (accu: Set<any>, selectedManagedIndex: { label: string; value?: ManagedIndexItem }) => {
        if (!selectedManagedIndex.value) return accu;
        const policy = selectedManagedIndex.value.policy;
        if (!policy) return accu;
        policy.states.forEach((state: State) => {
          accu.add(state.name);
        });
        return accu;
      },
      new Set()
    );

    const options = stateFilterSearchValue.trim() ? [{ label: `${stateFilterSearchValue}*` }] : [];
    const stateOptions = options.concat([...uniqueStates].map((stateName: string) => ({ label: stateName })));

    return (
      <ContentPanel bodyStyles={{ padding: "initial" }} title="Managed indices" titleSize="s">
        <div style={{ paddingLeft: "10px" }}>
          <EuiText size="xs">
            <p>
              When a managed index changes policy, the new policy will take affect after the managed index is finished running the current
              state or the new policy has the same structure as the current one.
            </p>
          </EuiText>
          <EuiSpacer size="m" />
          <EuiFormRow
            label="Managed indices"
            helpText="You can use a * as a wildcard in your index pattern"
            isInvalid={!!managedIndicesError}
            error={managedIndicesError}
          >
            <EuiComboBox
              placeholder="Choose managed indices"
              async
              options={managedIndices}
              isInvalid={!!managedIndicesError}
              selectedOptions={selectedManagedIndices}
              isLoading={managedIndicesIsLoading}
              // @ts-ignore
              onChange={this.props.onChangeManagedIndices}
              onSearchChange={this.onManagedIndexSearchChange}
            />
          </EuiFormRow>

          <EuiFormRow label="State filters" helpText="Apply new policy only on managed indices in these states.">
            <EuiComboBox
              placeholder="Choose state filters"
              options={stateOptions}
              selectedOptions={selectedStateFilters}
              // @ts-ignore
              onChange={this.props.onChangeStateFilters}
              onSearchChange={this.onStateFilterSearchChange}
            />
          </EuiFormRow>
        </div>
      </ContentPanel>
    );
  }
}
