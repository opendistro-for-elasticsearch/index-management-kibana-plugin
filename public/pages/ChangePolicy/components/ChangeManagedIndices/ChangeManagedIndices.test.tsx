/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, wait } from "@testing-library/react";
import ChangeManagedIndices from "./ChangeManagedIndices";
import { browserServicesMock } from "../../../../../test/mocks";
import coreServicesMock from "../../../../../test/mocks/coreServicesMock";
import { CoreServicesContext } from "../../../../components/core_services";

describe("<ChangeManagedIndices /> spec", () => {
  it("renders the component", async () => {
    browserServicesMock.managedIndexService.getManagedIndices = jest.fn().mockResolvedValue({ ok: true, response: { hits: { hits: [] } } });
    const { container } = render(
      <CoreServicesContext.Provider value={coreServicesMock}>
        <ChangeManagedIndices
          managedIndexService={browserServicesMock.managedIndexService}
          selectedManagedIndices={[]}
          selectedStateFilters={[]}
          onChangeManagedIndices={() => {}}
          onChangeStateFilters={() => {}}
          managedIndicesError=""
        />
      </CoreServicesContext.Provider>
    );

    await wait();

    expect(container.firstChild).toMatchSnapshot();
  });

  it("shows danger toaster when search fails", async () => {
    browserServicesMock.managedIndexService.getManagedIndices = jest.fn().mockRejectedValue(new Error("this is an error"));
    render(
      <CoreServicesContext.Provider value={coreServicesMock}>
        <ChangeManagedIndices
          managedIndexService={browserServicesMock.managedIndexService}
          selectedManagedIndices={[]}
          selectedStateFilters={[]}
          onChangeManagedIndices={() => {}}
          onChangeStateFilters={() => {}}
          managedIndicesError=""
        />
      </CoreServicesContext.Provider>
    );

    await wait();

    expect(coreServicesMock.notifications.toasts.addDanger).toHaveBeenCalledTimes(1);
    expect(coreServicesMock.notifications.toasts.addDanger).toHaveBeenCalledWith("this is an error");
  });

  it("shows danger toaster when search gracefully fails", async () => {
    browserServicesMock.managedIndexService.getManagedIndices = jest.fn().mockResolvedValue({ ok: false, error: "some error" });
    render(
      <CoreServicesContext.Provider value={coreServicesMock}>
        <ChangeManagedIndices
          managedIndexService={browserServicesMock.managedIndexService}
          selectedManagedIndices={[]}
          selectedStateFilters={[]}
          onChangeManagedIndices={() => {}}
          onChangeStateFilters={() => {}}
          managedIndicesError=""
        />
      </CoreServicesContext.Provider>
    );

    await wait();

    expect(coreServicesMock.notifications.toasts.addDanger).toHaveBeenCalledTimes(1);
    expect(coreServicesMock.notifications.toasts.addDanger).toHaveBeenCalledWith("some error");
  });

  it("shows danger toaster when search fails because of no config index", async () => {
    browserServicesMock.managedIndexService.getManagedIndices = jest
      .fn()
      .mockResolvedValue({ ok: false, error: "[index_not_found_exception]and other stuff" });
    render(
      <CoreServicesContext.Provider value={coreServicesMock}>
        <ChangeManagedIndices
          managedIndexService={browserServicesMock.managedIndexService}
          selectedManagedIndices={[]}
          selectedStateFilters={[]}
          onChangeManagedIndices={() => {}}
          onChangeStateFilters={() => {}}
          managedIndicesError=""
        />
      </CoreServicesContext.Provider>
    );

    await wait();

    expect(coreServicesMock.notifications.toasts.addDanger).toHaveBeenCalledTimes(1);
    expect(coreServicesMock.notifications.toasts.addDanger).toHaveBeenCalledWith("You have not created a managed index yet");
  });
});
