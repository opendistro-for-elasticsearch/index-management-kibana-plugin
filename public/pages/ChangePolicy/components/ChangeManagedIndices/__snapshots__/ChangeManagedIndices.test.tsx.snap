// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChangeManagedIndices /> spec renders the component 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium"
  style="padding-left: 0px; padding-right: 0px;"
>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
    style="padding: 0px 10px;"
  >
    <div
      class="euiFlexItem"
    >
      <h3
        class="euiTitle euiTitle--small"
      >
        Choose managed indices
      </h3>
    </div>
  </div>
  <hr
    class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
  />
  <div
    style="padding: initial;"
  >
    <div
      style="padding-left: 10px;"
    >
      <div
        class="euiSpacer euiSpacer--m"
      />
      <div
        class="euiFormRow"
        id="some_html_id-row"
      >
        <div
          class="euiFormRow__labelWrapper"
        >
          <label
            aria-invalid="false"
            class="euiFormLabel euiFormRow__label"
            for="some_html_id"
          >
            Managed indices
          </label>
        </div>
        <div
          class="euiFormRow__fieldWrapper"
        >
          <div
            aria-describedby="some_html_id-help"
            aria-expanded="false"
            aria-haspopup="listbox"
            class="euiComboBox"
            role="combobox"
          >
            <div
              class="euiFormControlLayout"
            >
              <div
                class="euiFormControlLayout__childrenWrapper"
              >
                <div
                  class="euiComboBox__inputWrap euiComboBox__inputWrap-isClearable"
                  data-test-subj="comboBoxInput"
                  tabindex="-1"
                >
                  <div
                    class="euiComboBox__input"
                    style="font-size: 14px; display: inline-block;"
                  >
                    <input
                      aria-controls=""
                      data-test-subj="comboBoxSearchInput"
                      id="some_html_id"
                      role="textbox"
                      style="box-sizing: content-box; width: 2px;"
                      value=""
                    />
                    <div
                      style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                    />
                  </div>
                </div>
                <div
                  class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                >
                  <button
                    aria-label="Open list of options"
                    class="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
                    data-test-subj="comboBoxToggleListButton"
                    type="button"
                  >
                    EuiIconMock
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFormHelpText euiFormRow__text"
            id="some_html_id-help"
          >
            You can use * as wildcards to form index patterns.
          </div>
        </div>
      </div>
      <div
        class="euiFormRow"
        id="some_html_id-row"
      >
        <div
          class="euiFormRow__labelWrapper"
        >
          <label
            class="euiFormLabel euiFormRow__label"
            for="some_html_id"
          >
            State filters
          </label>
        </div>
        <div
          class="euiFormRow__fieldWrapper"
        >
          <div
            aria-describedby="some_html_id-help"
            aria-expanded="false"
            aria-haspopup="listbox"
            class="euiComboBox euiComboBox-isDisabled"
            role="combobox"
          >
            <div
              class="euiFormControlLayout"
            >
              <div
                class="euiFormControlLayout__childrenWrapper"
              >
                <div
                  class="euiComboBox__inputWrap"
                  data-test-subj="comboBoxInput"
                  tabindex="-1"
                >
                  <p
                    class="euiComboBoxPlaceholder"
                  >
                    Choose state filters
                  </p>
                  <div
                    class="euiComboBox__input"
                    style="font-size: 14px; display: inline-block;"
                  >
                    <input
                      aria-controls=""
                      data-test-subj="comboBoxSearchInput"
                      disabled=""
                      id="some_html_id"
                      role="textbox"
                      style="box-sizing: content-box; width: 2px;"
                      value=""
                    />
                    <div
                      style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                    />
                  </div>
                </div>
                <div
                  class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                >
                  <button
                    aria-label="Open list of options"
                    class="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
                    data-test-subj="comboBoxToggleListButton"
                    disabled=""
                    type="button"
                  >
                    EuiIconMock
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFormHelpText euiFormRow__text"
            id="some_html_id-help"
          >
            Apply new policy only on managed indices in these states.
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
